addons/uom/static/src/js/uom_field_widget.js

```diff
@@ -0,0 +1,147 @@
+odoo.define('uom.UomFieldWidget', function (require) {
+"use strict";
+
+var core = require('web.core');
+var session = require('web.session');
+var basic_fields = require('web.basic_fields');
+var fieldUtils = require('web.field_utils');
+var field_registry = require('web.field_registry');
+var utils = require('web.utils');
+
+
+var _t = core._t;
+
+var FieldFloat = basic_fields.FieldFloat;
+var InputField = basic_fields.InputField
+
+var FieldUom = FieldFloat.extend({
+    supportedFieldTypes: ['float', 'uom'],
+    resetOnAnyFieldChange: true, // Have to listen to uom changes
+
+    events: _.extend({}, InputField.prototype.events, {
+        'input': '_onInput',
+    }),
+
+    /**
+     * Float fields using a uom widget have an additional uom_field
+     * parameter which defines the name of the field from which the uom
+     * should be read.
+     *
+     * If no uom field is given or the field does not exist, we fallback
+     * to the default input behavior instead.
+     *
+     * @override
+     */
+    init: function () {
+        this._super.apply(this, arguments);
+        this._setUom();
+        this.formatOptions.uom = this.uom;
+    },
+
+    /**
+     * Re-gets the uom as its value may have changed.
+     * @see FieldUom.resetOnAnyFieldChange
+     *
+     * @override
+     * @private
+     */
+    _reset: function () {
+        this._super.apply(this, arguments);
+        this._setUom();
+    },
+
+    /**
+     * Deduces the uom description from the field options and view state.
+     * The description is then available at this.uom.
+     *
+     * @private
+     */
+    _setUom: function () {
+        var uomField = this.nodeOptions.uom_field || this.field.uom_field || 'uom_id';
+        var uomID = this.record.data[uomField] && this.record.data[uomField].res_id;
+        this.uom = session.uoms[uomID];
+        this.formatOptions.uom = this.uom; // _formatValue() uses formatOptions
+    },
+
+    /**
+     * Allow to enter decimal digits base on the uom's rounding value.
+     *
+     * @override
+     * @private
+     */
+    _onInput: function() {
+        this._super.apply(this, arguments);
+        var value = this.$input.val();
+        var uom_precision = this.formatOptions.uom && this.formatOptions.uom.decimal_places;
+        if (value.includes('.')) {
+            var precision = value.split('.')[1].length;
+            if (uom_precision < precision) {
+                this.do_warn(_.str.sprintf(_t('Only %s decimals are allowed.</br> This setting can be configured on the Unit of Measure.'), uom_precision));
+                this.$input.val(value.slice(0,-1))
+            }
```

Comment:
do we really want to notify the user on input? I'd personally do that in 'change'